Userland/Libraries/LibWeb/Layout/BlockContainer.cpp

```diff
@@ -12,11 +13,19 @@ namespace Web::Layout {
 BlockContainer::BlockContainer(DOM::Document& document, DOM::Node* node, NonnullRefPtr<CSS::StyleProperties> style)
     : Box(document, node, move(style))
 {
+    if (dom_node() && dom_node()->is_html_input_element()) {
+        set_natural_width(CSS::Length(static_cast<HTML::HTMLInputElement*>(dom_node())->size(), CSS::Length::Type::Ch).to_px(*this));
+        set_natural_height(CSS::Length(1, CSS::Length::Type::Lh).to_px(*this));
+    }
```

Comment:
I don't think it's enough to do this when constructing a `BlockContainer`, since that means it won't get updated if you do stuff like `input.setAttribute('size', 1234)`

How about this: we put this logic in a helper on `HTMLInputElement`, something like `HTMLInputElement::update_natural_size()`, and then have it poke natural height values into the `HTMLInputElement`'s `layout_node()` instead?

Then we can call this in `create_layout_node()`, as well as in `attribute_changed()` when `size` changes.

Reply:
Ok, I'm trying to think this through. If the size is changed, then in `run_attribute_change_steps` we bump the DOM version but that doesn't, on its own, require re-layout. But don't we want to redo layout? Like, what if the `<input>` element's container is shrink-to-fit? Then the `<input>` element resizing will potentially affect the layout of a whole lot of stuff.

Reply:
I guess I'm proposing invalidating layout in `set_size`.

Reply:
Oh, of course! Yes, you are right we need to redo layout. :)

Reply:
Ok, added an `invalidate_layout` call in `set_size`. The `invalidate_layout`, instead of `set_needs_layout`, because the sizing is done on `BlockContainer` construction. It's possible to move that code into `BlockFormattingContext` and then not need to tear down the layout tree, but we looked and I think that code would be much more complicated & I don't think we need to make `set_size` super fast?

Reply:
Hmm that's still pretty awkward.

Given this:

- The layout nodes that correspond to replaced elements currently have a `prepare_for_replaced_layout()` virtual function. We use this function to tell the layout node that it's time to figure out its own natural dimensions and aspect ratio.

- This PR is introducing a new behavior where input elements also behave as replaced elements during layout, specifically in order for natural size to be used.

How about this:

- We move the `prepare_for_replaced_layout()` from `Layout::ReplacedBox` up to `Layout::Box`, so that we can call it on `BlockContainer` as well.

- We make sure that anyone who was calling `prepare_for_replaced_layout()` no longer checks `is<ReplacedBox>()`, but simply that `box_is_sized_as_replaced_element()`.

- We move the input element size derivation out of `BlockContainer`'s constructor, and into `BlockContainer::prepare_for_replaced_layout()`.

Reply:
This might be causing issues.  I implemented it and some test are now failing that weren't before.  I'm looking into it.